### PR TITLE
[firebase_auth_platform_interface] Add AppleAuthCredential class

### PR DESCRIPTION
**Pull request by zariweyo**
## Description
This PR add AppleAuthCredential class to use as requirement in the nest PR to solve Apple sign in firebase for iOS devices

## Related Issues
This PR resolve this issue in firebase_auth: Sign in with Apple #1434 for iOS.
https://github.com/FirebaseExtended/flutterfire/issues/1434

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_auth/firebase_auth_platform_interface/lib/src/types.dart [hunk 1]
```diff
@@ -159,6 +159,28 @@ class GoogleAuthCredential extends AuthCredential {
...
+/// An [AuthCredential] for authenticating via apple.com.
+class AppleAuthCredential extends AuthCredential {
+  const AppleAuthCredential({
+    @required this.idToken,
+    @required this.accessToken,
+  }) : super(_providerId);
+
+  static const String _providerId = 'apple.com';
+
+  /// The Google ID token.
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth_platform_interface/lib/src/types.dart]:
```suggestion
  /// The Apple ID token.
```

File: packages/firebase_auth/firebase_auth_platform_interface/lib/src/types.dart [hunk 2]
```diff
@@ -159,6 +159,28 @@ class GoogleAuthCredential extends AuthCredential {
...
+  final String idToken;
+
+  /// The Google access token.
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth_platform_interface/lib/src/types.dart]:
```suggestion
  /// The Apple access token.
```

File: packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ author: Flutter Team <flutter-dev@googlegroups.com>
 homepage: https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_auth/firebase_auth_platform_interface
 # NOTE: We strongly prefer non-breaking changes, even at the expense of a
 # less-clean API. See https://flutter.dev/go/platform-interface-breaking-changes
-version: 1.0.0
+version: 1.0.1
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml]: This should be 1.1.0

File: packages/firebase_auth/firebase_auth_platform_interface/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 1.0.1
+
+- Added type AppleAuthCredential.
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth_platform_interface/CHANGELOG.md]:
```suggestion
- Added type `AppleAuthCredential`.
```

**Comment by hterkelsen** [comment]
Do you have a work-in-progress PR for supporting this in Android and iOS? From the Firebase guides, it seems they use the generic `OAuthProvider` and `OAuthCredential` to achieve this. I'm fine with having a special-case `AppleAuthCredential` for this, though.

**Comment by zariweyo** [comment]
@hterkelsen yes, I have the PR https://github.com/FirebaseExtended/flutterfire/pull/1526 waiting for this. And I need this PR to solve that. Thanks all.

File: packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml (hunk 3 above)
Comment (zariweyo) [reply packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml]: Ok, I accept all your suggestions and change version to 1.1.0.
Thanks

**Comment by zariweyo** [comment]
@collinjackson Don't worry for extra work, but I'm really not satisfaced with this solution because, what is the correct value for _providerId?, I think that it's a mistake to use a generic auth provider, like OauthProvider, to a specific solution for Apple.

**Comment by collinjackson** [comment]
I would put both of these classes in types.h:

```
class OAuthProvider {
  OAuthProvider({ @required this.providerId });

  final String providerId;

  static AuthCredential getCredential({
    @required String idToken,
    @required String accessToken,
    String secret,
  }) {
    return OAuthCredential._(idToken: idToken, accessToken: accessToken, providerId: providerId, secret: secret);
  }
}

class OAuthCredential extends AuthCredential {
  const OAuthCredential._({
    @required this.idToken,
    @required this.accessToken,
    @required this.providerId,
    this.secret;
  }) : super(providerId);

  /// The ID Token associated with this credential.
  final String idToken;

  /// The OAuth access token.
  final String accessToken;

    /// The OAuth access token.
  final String secret;

  @override
  Map<String, String> _asMap() => <String, String>{
        'idToken': idToken,
        'accessToken': accessToken,
        'providerId': providerId,
        'secret': secret,
      };
}
```

**Comment by collinjackson** [comment]
I ran the above approach past @hterkelsen and he agreed.

**Comment by zariweyo** [comment]
I have changed your proposal a little and probe in the future implementation of firebase_auth, that will be a generic provider.

File: packages/firebase_auth/firebase_auth_platform_interface/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+## 1.1.0
+
+- Added type `PlatformOAuthCredential` for generic Oauth providers.
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth_platform_interface/CHANGELOG.md]:
```suggestion
- Added type `PlatformOAuthCredential` for generic OAuth providers.
```

**Comment by Ahmadre** [comment]
@hterkelsen any news?

**Pull request merged**